### ISSUE TITLE
Volume attached to a node outside of kubernetes cluster

### ISSUE THREAD

**Issue by seimic** [hetznercloud/csi-driver#107]
Hi,

maybe I miss some settings but how to force the driver to attach volumes to specific nodes?

I have had now a very confusing result in following setup (simplified example here):

- In hcloud there are 7 nodes with following names (edge, n01, n02, n03, n04, n05, n06).
- Kubernetes uses only the n01-n06 nodes. Edge is outside of kubernetes and the only hcloud node with active public interface. All others are using local networking and are registered with their local IPs only.
- Nodes n01-n03 have additional taints and annotations in kubernetes to prevent pods/deployments to be scheduled on them
- In the csi-driver deployment this taints and annotations are used to deploy the driver only on this 3 nodes

```
affinity:
  nodeAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
	  nodeSelectorTerms:
	  - matchExpressions:
	    - key: node-role.example.com/data
          operator: In
		  values: 
		    - "true"
tolerations:
  - key: "node-role.example.com/data"
    operator: "Equal"
    value: "true"
    effect: NoSchedule
```
Result of creating a volume (mariadb helm chart with the hcloudvolume class) is - the volume is attached to the "edge" server which is not part of the kubernetes clusters. Its not accessible.

I was expecting it to create and attach the volumes to the hcloud-csi-nodes, where the driver is installed (n01-n03). 

I try to dedicate few nodes to storage and collocate database on it. Only the database should run at those nodes without any other pods/deployments.

Is this somehow possible?

Kind regards,
Michael

**Comment by seimic** [comment]
OK found the reason...
You resolve the instance id by a request. Probably using the caller IP or hostname from the request header (Masqueraded in the edge node). The traffic passes the edge node (there are two in a HA setup with floating IP) which is the only way in/out of the private network. 

```
func getInstanceID() (int, error) {
	resp, err := http.Get("http://169.xxx.xxx.xxx/2009-04-04/meta-data/instance-id")
        ...
```

I was thinking you resolve the instance id using the hcloud api and host names like this
`curl -H "Authorization: Bearer $API_TOKEN" "https://api.hetzner.cloud/v1/servers?name=$(hostname)"`
Costs a bit more to parse the results (csi-node could do when initialising) but you get what you ask for in this case.

In other words I have to mount the volumes on my own in terraform and use a local path provisioner.

Kind regards,
Michael

**Comment by seimic** [comment]
OK this in cloudinit (debian/ubuntu) should do the trick too

`echo "export HCLOUD_SERVER_ID=$(curl -sL -H "Authorization: Bearer $API_TOKEN" https://api.hetzner.cloud/v1/servers?name=$(hostname) | jq -r '.servers[0] | .id')" >> /etc/environment`

for centos /etc/profile.env

**Comment by seimic** [comment]
Unfortunately this didn't work. Changing the node-ids in the node annotations has had no effect too. I suppose this is read only once at some initialization step. I'm not sure which pods need the node-id environment variable except the driver itself.

**Comment by LKaemmerling** [comment]
The controller itself is in favor of attaching the nodes. There is currently no other way than using the metadata service (https://docs.hetzner.cloud/#overview-server-metadata). 

The recommended way (in your setup) is that you route this IP(169.254.169.254) over the public interface of the server. If you disable the public interface of a node completely all functionally that uses the metadata service (like cloud-init) does not work anymore.
